server/statistics/hot_cache.go

```diff
@@ -47,178 +50,319 @@ const (
 	ReadFlow
 )
 
-// HotSpotCache is a cache hold hot regions.
-type HotSpotCache struct {
-	writeFlow cache.Cache
-	readFlow  cache.Cache
+// HotStoresStats saves the hotspot peer's statistics.
+type HotStoresStats struct {
+	hotStoreStats  map[uint64]cache.Cache         // storeID -> hot regions
+	storesOfRegion map[uint64]map[uint64]struct{} // regionID -> storeIDs
 }
 
-// NewHotSpotCache creates a new hot spot cache.
-func NewHotSpotCache() *HotSpotCache {
-	return &HotSpotCache{
-		writeFlow: cache.NewCache(statCacheMaxLen, cache.TwoQueueCache),
-		readFlow:  cache.NewCache(statCacheMaxLen, cache.TwoQueueCache),
+// NewHotStoresStats creates a HotStoresStats
+func NewHotStoresStats() *HotStoresStats {
+	return &HotStoresStats{
+		hotStoreStats:  make(map[uint64]cache.Cache),
+		storesOfRegion: make(map[uint64]map[uint64]struct{}),
 	}
 }
 
-// CheckWrite checks the write status, returns whether need update statistics and item.
-func (w *HotSpotCache) CheckWrite(region *core.RegionInfo, stats *StoresStats) (bool, *RegionStat) {
+// CheckRegionFlow checks the flow information of region.
+func (f *HotStoresStats) CheckRegionFlow(region *core.RegionInfo, kind FlowKind) []HotSpotPeerStatGenerator {
 	var (
-		WrittenBytesPerSec uint64
-		value              *RegionStat
+		generators    []HotSpotPeerStatGenerator
+		getBytesFlow  func() uint64
+		getKeysFlow   func() uint64
+		bytesPerSec   uint64
+		keysPerSec    uint64
+		oldRegionStat *HotSpotPeerStat
 	)
 
-	WrittenBytesPerSec = uint64(float64(region.GetBytesWritten()) / float64(RegionHeartBeatReportInterval))
+	storeIDs := make(map[uint64]struct{})
+	// got the storeIDs, including old region and new region
+	ids, ok := f.storesOfRegion[region.GetID()]
+	if ok {
+		for storeID := range ids {
+			storeIDs[storeID] = struct{}{}
+		}
 
-	v, isExist := w.writeFlow.Peek(region.GetID())
-	if isExist {
-		value = v.(*RegionStat)
-		// This is used for the simulator.
-		if !Simulating {
-			interval := time.Since(value.LastUpdateTime).Seconds()
-			if interval < minHotRegionReportInterval {
-				return false, nil
-			}
-			WrittenBytesPerSec = uint64(float64(region.GetBytesWritten()) / interval)
+	}
+	for _, peer := range region.GetPeers() {
+		// ReadFlow no need consider the followers.
+		if kind == ReadFlow && peer.GetStoreId() != region.GetLeader().GetStoreId() {
+			continue
+		}
+		if _, ok := storeIDs[peer.GetStoreId()]; !ok {
+			storeIDs[peer.GetStoreId()] = struct{}{}
```

Comment:
We should consider the peers of the region has been changed, such as from store id [1, 2, 3] to [1, 3, 4].

Reply:
Then when will you update `f`? It's confusing to use these stores without update `f.storesOfRegion`. If `f` will be updated in other codes, I think we should make sure those codes will run before this function, instead of consider this situation here.

Reply:
If we got `update items`, we will update it. `Check` is used to checks if there are some items need an update or delete, which only read the cache. we will update it later with a `Lock`.